File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.js
const preorderTraversal = (root) => {
    const values = [];
    const stack = [];

    let current = root;

    while (current !== null || stack.length !== 0) {
        if (current !== null) {
            values.push(current.val);
            stack.push(current);
            current = current.left;
        } else {
            const last = stack.pop();
            current = last.right;
        }
    }

    return values;
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        values = []
        stack = []

        current = root

        while current is not None or len(stack) > 0:
            if current is not None:
                values.append(current.val)
                stack.append(current)
                current = current.left
            else:
                last = stack.pop()
                current = last.right
        
        return values
